### PR TITLE
Azure Front Door Rules Engine commands preview

### PR DESCRIPTION
**Pull request by aramesh-msft**
* add rules-engine commands
* upgrade SDK (swagger) version to 2020-01-01 for front-door

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S167

**Comment by aramesh-msft** [comment]
@johsun-msft is added to the review. #Closed

**Comment by aramesh-msft** [comment]
If I remove "universal=1" from setup.cfg then that will be blank, is that the intention or remove the file completely?

**Comment by aramesh-msft** [comment]
> https://github.com/Azure/azure-cli-extensions/blob/d5baa785ce337b3334fa0c070c509dbe2e532120/src/front-door/setup.cfg#L1-L2
> 
There is some discussion here https://github.com/Azure/azure-cli-extensions/pull/1441 I want to make

**Comment by haroldrandom** [comment]
> > https://github.com/Azure/azure-cli-extensions/blob/d5baa785ce337b3334fa0c070c509dbe2e532120/src/front-door/setup.cfg#L1-L2
> 
> There is some discussion here #1441 I want to make sure removing these 2 lines wouldn't cause any side-effects.

okay, then keep the setup.cfg untouched.

**Comment by aramesh-msft** [comment]
@MyronFanQiu Could you please review and merge this PR? I don't have permissions to merge

File: azure-cli-extensions.pyproj [hunk 1]
```diff
@@ -2468,120 +2468,26 @@
     <Compile Include="src\find\setup.py" />
     <Compile Include="src\front-door\azext_front_door\commands.py" />
     <Compile Include="src\front-door\azext_front_door\custom.py" />
-    <Compile Include="src\front-door\azext_front_door\vendored_sdks\front_door_management_client.py" />
```
Comment (MyronFanQiu) [inline azure-cli-extensions.pyproj]: could we remove this file?

File: src/front-door/azext_front_door/commands.py [hunk 2]
```diff
@@ -170,3 +175,28 @@ def load_command_table(self, _):
...
+
+    with self.command_group('network front-door rules-engine', rules_engine_sdk, is_preview=True) as g:
+        g.show_command('show', 'get')
+        g.command('list', 'list_by_front_door')
+        g.command('delete', 'delete')
+
+    with self.command_group('network front-door rules-engine rule', rules_engine_sdk, is_preview=True) as g:
+        g.custom_command('create', 'create_rules_engine_rule')
+        g.custom_command('delete', 'delete_rules_engine_rule')
+        g.custom_command('show', 'show_rules_engine_rule')
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/commands.py]: please use custom_show_command

File: src/front-door/azext_front_door/custom.py [hunk 3]
```diff
@@ -519,14 +515,30 @@ def create_fd_routing_rules(cmd, resource_group_name, front_door_name, item_name
...
         from knack.util import CLIError
         raise CLIError(forwarding_usage)
-    if route_type == 'Redirect' and any([custom_forwarding_path, forwarding_protocol, backend_pool,
+    if 'Redirect' in route_type and any([custom_forwarding_path, forwarding_protocol, backend_pool,
                                          query_parameter_strip_directive, dynamic_compression]):
         from knack.util import CLIError
         raise CLIError(redirect_usage)
 
+
+def create_fd_routing_rules(cmd, resource_group_name, front_door_name, item_name, frontend_endpoints, route_type,
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/custom.py]: shoud create method support rules_engine too?

File: src/front-door/azext_front_door/custom.py [hunk 4]
```diff
@@ -1098,3 +1112,301 @@ def list_custom_rule_match_conditions(cmd, resource_group_name, policy_name, rul
...
+        raise CLIError("Rules Engine must at least contain one rule")
+
+    return client.create_or_update(resource_group_name, front_door_name, rules_engine_name, rules=rules_engine.rules)
+
+
+def show_rules_engine_rule(cmd, resource_group_name, front_door_name, rules_engine_name, rule_name):
+    client = cf_fd_rules_engines(cmd.cli_ctx, None)
+    rules_engine = client.get(resource_group_name, front_door_name, rules_engine_name)
+    try:
+        return next(x for x in rules_engine.rules if x.name.lower() == rule_name.lower())
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/custom.py]: I'm not sure whether we can return a iterator or not. Could you please add a test for the show command? or we can turn it by `list(next())`

File: src/front-door/azext_front_door/custom.py [hunk 5]
```diff
@@ -1098,3 +1112,301 @@ def list_custom_rule_match_conditions(cmd, resource_group_name, policy_name, rul
...
+
+
+def add_rules_engine_condition(cmd, resource_group_name, front_door_name, rules_engine_name,
+                               rule_name, match_variable=None, operator=None, match_values=None,
+                               selector=None, negate_condition=None, transforms=None):
+    from azext_front_door.vendored_sdks.models import RulesEngineMatchCondition
+    client = cf_fd_rules_engines(cmd.cli_ctx, None)
+    rules_engine = client.get(resource_group_name, front_door_name, rules_engine_name)
+
+    foundRule = False
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/custom.py]: please use found_rule

File: src/front-door/azext_front_door/tests/latest/test_route_rule_scenarios.py [hunk 6]
```diff
@@ -22,7 +22,7 @@ def test_route_rule_basic(self, resource_group):
                  '--patterns /forward2 --caching Disabled',
                  checks=[
                      self.check('patternsToMatch[0]', '/forward2'),
-                     self.check('contains(keys(routeConfiguration), "cacheConfiguration")', False)
+                     self.check('routeConfiguration.cacheConfiguration', None)
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/tests/latest/test_route_rule_scenarios.py]: is there any reason to change it?

File: src/front-door/azext_front_door/_params.py [hunk 7]
```diff
@@ -263,3 +266,59 @@ def load_arguments(self, _):
         c.argument('rule_name', options_list=['--name', '-n'], help='Name of the custom rule.')
         c.argument('policy_name', waf_policy_name_type, id_part=None)
     # endregion
+
+    # region RulesEngine
+    with self.argument_context('network front-door rules-engine rule') as c:
+        c.argument('front_door_name', frontdoor_name_type, options_list=['--front-door-name', '-n'])
+        c.argument('rules_engine_name', rules_engine_name_type, id_part=None)
+        c.argument('rule_name', help='Name of the rule')
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/_params.py]: Usually in Azure CLI, the deepest name should support `-n` and `--name`. The `front_door_name` should be just `--front-door-name`

File: src/front-door/azext_front_door/_params.py [hunk 8]
```diff
@@ -263,3 +266,59 @@ def load_arguments(self, _):
...
+        c.argument('action_type', arg_group="Action", arg_type=get_enum_type(['RequestHeader', 'ResponseHeader', 'ForwardRouteOverride', 'RedirectRouteOverride']), help='Action type to apply for a rule.')
+        c.argument('header_action', arg_group="Action", arg_type=get_enum_type(HeaderActionType), help='Header action type for the requests.')
+        c.argument('header_name', arg_group="Action", help='Name of the header to modify.')
+        c.argument('header_value', arg_group="Action", help='Value of the header.')
+        c.argument('match_variable', arg_group="Match Condition", arg_type=get_enum_type(RulesEngineMatchVariable), help='Name of the match condition.')
+        c.argument('operator', arg_group="Match Condition", arg_type=get_enum_type(RulesEngineOperator), help='Operator of the match condition.')
+        c.argument('match_values', arg_group="Match Condition", nargs='+', help='Space-separated list of values to match against.')
+        c.argument('selector', arg_group="Match Condition", help='Optional selector for the match condition variable.')
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/_params.py]: should it be an enum? what are the valid values?

File: src/front-door/azext_front_door/_params.py [hunk 9]
```diff
@@ -263,3 +266,59 @@ def load_arguments(self, _):
...
+        c.argument('backend_pool', help='Name or ID of a backend pool.', validator=validate_backend_pool)
+        c.argument('forwarding_protocol', arg_type=get_enum_type(FrontDoorForwardingProtocol), help='Protocol to use for forwarding traffic.')
+        c.argument('custom_forwarding_path', help='Custom path used to rewrite resource paths matched by this rule. Leave empty to use incoming path.')
+        c.argument('caching', arg_type=get_three_state_flag(positive_label='Enabled', negative_label='Disabled', return_label=False), help='Whether to enable caching for this route.')
+        c.argument('cache_duration', help='The duration for which the content needs to be cached. Allowed format is ISO 8601 duration')
+        c.argument('dynamic_compression', arg_type=get_three_state_flag(positive_label='Enabled', negative_label='Disabled', return_label=True), help='Use dynamic compression for cached content.')
+        c.argument('query_parameter_strip_directive', arg_type=get_enum_type(FrontDoorQuery), help='Treatment of URL query terms when forming the cache key.')
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/_params.py]: this argument is quite long. Is `query_parameter` necessary?

File: src/front-door/azext_front_door/_params.py [hunk 10]
```diff
@@ -263,3 +266,59 @@ def load_arguments(self, _):
...
+        c.argument('query_parameters', help='Query parameters to include or exclude (comma separated).')
```
Comment (MyronFanQiu) [inline src/front-door/azext_front_door/_params.py]: a bit confused for me. How to include and how to exclue?

File: azure-cli-extensions.pyproj (hunk 1 above)
Comment (aramesh-msft) [reply azure-cli-extensions.pyproj]: Yes, the new vendored SDK directory structure has the same file as '_front_door_management_client.py'

---
In reply to: [398370729](https://github.com/Azure/azure-cli-extensions/pull/1432#discussion_r398370729) [](ancestors = 398370729)

File: src/front-door/azext_front_door/custom.py (hunk 3 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/custom.py]: Yeah, I guess you are right. Rules engine has to be created first and then linked to a route.  'create' word in function name threw me off, hence I added it only for the update routing rule method. In this case we can actually create a new route in a front door that can link an existing rules engine, I'll add it.

File: src/front-door/azext_front_door/tests/latest/test_route_rule_scenarios.py (hunk 6 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/tests/latest/test_route_rule_scenarios.py]: When I re-ran this test, ```keys(routeConfiguration)``` was returning 'None'. Hence ```contains(keys(routeConfiguration), "cacheConfiguration")``` would be returning False regardless.
```routeConfiguration.cacheConfiguration``` check was correctly testing for cache being disabled

File: src/front-door/azext_front_door/_params.py (hunk 7 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py]: hmm.. do you mean the --rules-engine-name should be -n instead?

File: src/front-door/azext_front_door/_params.py (hunk 8 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py]: No, selector is a [string](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/frontdoor/resource-manager/Microsoft.Network/stable/2020-01-01/frontdoor.json#L1921)

File: src/front-door/azext_front_door/_params.py (hunk 9 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py]: Yes, query parameter is a comma separated list that is used for changing [cache behavior](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/frontdoor/resource-manager/Microsoft.Network/stable/2020-01-01/frontdoor.json#L1565).

File: src/front-door/azext_front_door/_params.py (hunk 10 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py]: The list of values for example ['key1=value1', 'key2=value2'] will be included or excluded depending on the [query_parameter_strip_directive ](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/frontdoor/resource-manager/Microsoft.Network/stable/2020-01-01/frontdoor.json#L1569) chosen - "StripAllExcept" or "StripOnly" respectively.

File: src/front-door/azext_front_door/_params.py (hunk 8 above)
Comment (MyronFanQiu) [reply src/front-door/azext_front_door/_params.py]: got it

File: src/front-door/azext_front_door/_params.py (hunk 7 above)
Comment (MyronFanQiu) [reply src/front-door/azext_front_door/_params.py]: `rule_name` should be `--name/-n`

File: src/front-door/azext_front_door/_params.py (hunk 9 above)
Comment (MyronFanQiu) [reply src/front-door/azext_front_door/_params.py]: sure. I mean `--query-parameter-strip-directive` is too long. Is there any possible to shorten it?

File: src/front-door/azext_front_door/_params.py (hunk 10 above)
Comment (MyronFanQiu) [reply src/front-door/azext_front_door/_params.py]: could you add this relationship between this two arguments into the help message?

File: src/front-door/azext_front_door/commands.py (hunk 2 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/commands.py]: done in next commit

File: src/front-door/azext_front_door/custom.py (hunk 3 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/custom.py]: done in next commit

File: src/front-door/azext_front_door/custom.py (hunk 4 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/custom.py]: Added test case, it works fine and so is [this line](https://github.com/Azure/azure-cli-extensions/blob/master/src/front-door/azext_front_door/custom.py#L1028) above.

File: src/front-door/azext_front_door/_params.py (hunk 7 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py]: done params changes in next commit

File: src/front-door/azext_front_door/_params.py (hunk 9 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py]: We are already using query-parameter-strip-directive in [create_fd_routing_rule](https://github.com/Azure/azure-cli-extensions/blob/master/src/front-door/azext_front_door/custom.py#L504), I don't want to break any existing scripts that folks may have. what do you think?

File: src/front-door/azext_front_door/_params.py (hunk 10 above)
Comment (aramesh-msft) [reply src/front-door/azext_front_door/_params.py]: Yes, good point, added help text in next commit

**Pull request merged**